pkg/controllers/management/auth/role_template_handler.go

```diff
@@ -1,47 +1,105 @@
+	roleTemplateHandlerName = "mgmt-auth-roletemplate-handler"
 	prtbByRoleTemplateIndex = "management.cattle.io/prtb-by-role-template"
 )
 
 type roleTemplateHandler struct {
-	prtbIndexer cache.Indexer
-	prtbClient  v3.ProjectRoleTemplateBindingInterface
+	prtbIndexer    cache.Indexer
+	prtbClient     v3.ProjectRoleTemplateBindingInterface
+	clusters       v3.ClusterInterface
+	clusterManager *clustermanager.Manager
 }
 
-func newRoleTemplateHandler(management *config.ManagementContext) *roleTemplateHandler {
+func newRoleTemplateHandler(management *config.ManagementContext, clusterManager *clustermanager.Manager) v3.RoleTemplateLifecycle {
 	informer := management.Management.ProjectRoleTemplateBindings("").Controller().Informer()
 	indexers := map[string]cache.IndexFunc{
 		prtbByRoleTemplateIndex: prtbByRoleTemplate,
 	}
 	informer.AddIndexers(indexers)
 
 	rtl := &roleTemplateHandler{
-		prtbIndexer: informer.GetIndexer(),
-		prtbClient:  management.Management.ProjectRoleTemplateBindings(""),
+		prtbIndexer:    informer.GetIndexer(),
+		prtbClient:     management.Management.ProjectRoleTemplateBindings(""),
+		clusters:       management.Management.Clusters(""),
+		clusterManager: clusterManager,
 	}
 	return rtl
 }
 
-// sync is called periodically and on real updates
-func (rtl *roleTemplateHandler) sync(key string, obj *v3.RoleTemplate) (runtime.Object, error) {
-	if obj == nil || obj.DeletionTimestamp != nil {
-		return nil, nil
+func (rtl *roleTemplateHandler) Create(obj *v3.RoleTemplate) (runtime.Object, error) {
+	if err := rtl.enqueuePrtbs(obj); err != nil {
+		return nil, err
 	}
+	return nil, nil
+}
+
+func (rtl *roleTemplateHandler) Updated(obj *v3.RoleTemplate) (runtime.Object, error) {
 	if err := rtl.enqueuePrtbs(obj); err != nil {
 		return nil, err
 	}
 	return nil, nil
 }
 
+func (rtl *roleTemplateHandler) Remove(obj *v3.RoleTemplate) (runtime.Object, error) {
+	clusters, err := rtl.clusters.List(metav1.ListOptions{})
+	if err != nil {
+		return obj, err
+	}
+
+	// Collect all the errors to delete as many user context cluster roles as possible
+	var allErrors []error
+
+	for _, cluster := range clusters.Items {
+		userContext, err := rtl.clusterManager.UserContext(cluster.Name)
+		if err != nil {
+			// ClusterUnavailable error indicates the record can't talk to the downstream cluster
+			if !IsClusterUnavailable(err) {
+				allErrors = append(allErrors, err)
+			}
+			continue
+		}
+
+		b, err := userContext.RBAC.ClusterRoles("").Controller().Lister().Get("", obj.Name)
+		if err != nil {
+			// User context clusterRole doesn't exist
+			if !apierrors.IsNotFound(err) {
+				allErrors = append(allErrors, err)
+			}
+			continue
+		}
+
+		err = userContext.RBAC.ClusterRoles("").Delete(b.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			// User context clusterRole doesn't exist
+			if !apierrors.IsNotFound(err) {
+				allErrors = append(allErrors, err)
+			}
+			continue
```

Comment:
nit: probably don't need continue

Reply:
It's not technically necessary but it's there to be clear about intentions and if something is added below this it won't cause a bug. grb-sync follows a similar pattern.